### DISCUSSION TITLE
Add Redis as a session storage option

### DISCUSSION THREAD

**Discussion by aleksasiriski** [dani-garcia/vaultwarden#3143] [Ideas]
This would allow multiple replicas to be spawned, which allows for HA in Kubernetes. Postgresql support is already here for user's vaults but sessions are still stateful.

**Comment by BlackDex** [comment]
Vaultwarden doesn't have any sessions. So, not very useful to add something like this.

Besides that, Vaultwarden can sorta run in HA. But there are a few items to thing about.
That is
- The `rsa_key.*` files need to be shared amongst the instances.
- Attachments and icon cache needs to be shared.
  Basically it means have a shared `/data` directory
- SQLite should not be used, only MariaDB/MySQL or PostgreSQL will work.

The only quirk that i know of is WebSockets which do not trigger other running instances of Vaultwarden to send the same message, since Vaultwarden doesn't know of the existence of the other running instances.

**Reply by aleksasiriski** [reply]
Oh, thanks. I just thought it saves sessions in a file since when I had 3 replicas I couldn't log in with the error "Your session expired!"

**Reply by BlackDex** [reply]
That is because of the `rsa_key` files which are all different. They cause the jwt tokens to be different and thus invalid per instance. If you make them the same via configmap or something, that should be enough.

But then stil attachments need to be shared though

**Reply by aleksasiriski** [reply]
That makes sense, I'll setup a secret and mount it as file for `rsa_key` and use Longhorn RWX for `/data`.

**Reply by tessus** [reply]
> The only quirk that i know of is WebSockets which do not trigger other running instances of Vaultwarden to send the same message, since Vaultwarden doesn't know of the existence of the other running instances.

You can workaround this by using stickyness on the LB. And in case a node fails, it's the same as connecting to vw again. All good.

**Reply by BlackDex** [reply]
That doesn't solve it. If your browser extension is connected to server 1, and your web-vault/browser is served by server 2. Now you update a vault item via the web-vault, the extension never receives the update, since server 2 didn't know that connection existed.

There are some ways to get this working, but that needs changes on the server side. Stuff like an internal websockets connection between the servers for example, or something like a gossip protocol between the servers. But that's going to be a lot of work. And i can think of other items to pickup first.

**Reply by tessus** [reply]
Yes, this is true, but I was rather thinking along the lines that it's rather likely that the browser extension and web-vault are used on the same machine (thus the same server). Even the mobile app would use the same server when connected to the same WiFi. 
But I see your point. In the scenario you painted stickiness won't solve the issue.
